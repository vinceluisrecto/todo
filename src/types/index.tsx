export interface Todo {
  id?: string;
  status: 'Pending' | 'In Progress' | 'Completed';
  title: string;
  description: string;
  due_date: string;
}

export interface AppContextType {
  todo: Todo[];
  addTodo: (todo: Todo) => void;
  deleteTodo: (id: string) => void;
  updateTodo: (todo: Todo) => void;
}
