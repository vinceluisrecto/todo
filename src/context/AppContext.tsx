import { createContext, useContext, useState, ReactNode } from 'react';
import { AppContextType, Todo } from '../types';

const AppContext = createContext<AppContextType | undefined>(undefined);

interface AppProviderProps {
  children: ReactNode;
}

export const AppProvider = ({ children }: AppProviderProps) => {
  const [todo, setTodo] = useState<Todo[]>([]);

  const addTodo = (todo: Todo) => {
    setTodo((props) => [...props, todo]);
  };

  const deleteTodo = (id: string) => {
    setTodo((props) => props.filter((p) => p.id !== id));
  };

  const updateTodo = (todo: Todo) => {
    setTodo((props) =>
      props.map((p) => {
        if (p.id === todo.id) return todo;
        return p;
      })
    );
  };

  return (
    <AppContext.Provider value={{ todo, addTodo, deleteTodo, updateTodo }}>
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = (): AppContextType => {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error('useAppContext must be used within an AppProvider');
  }
  return context;
};
