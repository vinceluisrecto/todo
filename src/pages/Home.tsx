import { Link } from 'react-router-dom';
import TodoList from '../components/TodoList';
import { useAppContext } from '../context/AppContext';
import { useState } from 'react';

const Home = () => {
  const { todo, deleteTodo } = useAppContext();
  const [status, setStatus] = useState<string>('All');
  const [search, setSearch] = useState<string>('');

  const filteredTodo = () => {
    const filteredTodo = todo.filter((t) => t.title.includes(search));
    if (status !== 'All')
      return filteredTodo.filter((t) => t.status === status);
    return filteredTodo;
  };

  const onStatusChange = (
    e: React.FormEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    setStatus(e.currentTarget.value);
  };

  const onSearchChange = (
    e: React.FormEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    setSearch(e.currentTarget.value);
  };

  return (
    <div className="min-w-full min-h-[100vh] flex items-center justify-center">
      <div className="my-5 flex flex-col items-center justify-center w-full max-w-[35%] border border-gray-300 rounded-lg p-6 gap-3">
        <h1 className="text-[29px]">To-Do List</h1>

        <div className="flex flex-row items-center justify-between w-full gap-3">
          <div className="flex-1">
            <label
              htmlFor="search"
              className="block mb-1 text-sm font-medium text-gray-900"
            >
              Search
            </label>
            <input
              id="search"
              type="text"
              name="title"
              className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
              placeholder="Title"
              onChange={onSearchChange}
              value={search}
            />
          </div>

          <div>
            <label
              htmlFor="status"
              className="block mb-1 text-sm font-medium text-gray-900"
            >
              Status
            </label>
            <select
              id="status"
              name="status"
              className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
              onChange={onStatusChange}
              value={status}
            >
              <option value="All">All</option>
              <option value="Pending">Pending</option>
              <option value="In Progress">In Progress</option>
              <option value="Completed">Completed</option>
            </select>
          </div>
        </div>

        <div className="w-full">
          <TodoList todos={filteredTodo()} deleteTodo={deleteTodo} />
        </div>

        <Link
          to="/add-todo"
          className="focus:outline-none text-white bg-green-700 hover:bg-green-800 focus:ring-4 focus:ring-green-300 font-medium rounded-lg text-sm px-5 py-2.5 me-2 mt-3 w-full text-center"
          data-testid="add-todo-button"
        >
          Add Todo
        </Link>
      </div>
    </div>
  );
};

export default Home;
