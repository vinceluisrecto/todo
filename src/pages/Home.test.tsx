import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import Home from './Home';
import { useAppContext } from '../context/AppContext';
import { Todo } from '../types';

// Mock the useAppContext to simulate the todo list and delete function
jest.mock('../context/AppContext', () => ({
  useAppContext: jest.fn(),
}));

describe('Home Component', () => {
  const mockDeleteTodo = jest.fn();
  const mockTodoList: Todo[] = [
    {
      id: '1',
      title: 'First Todo',
      description: 'First Description',
      status: 'Pending',
      due_date: '2024-11-21',
    },
    {
      id: '2',
      title: 'Second Todo',
      description: 'Second Description',
      status: 'In Progress',
      due_date: '2024-11-22',
    },
    {
      id: '3',
      title: 'Completed Todo',
      description: 'Completed Description',
      status: 'Completed',
      due_date: '2024-11-23',
    },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
    (useAppContext as jest.Mock).mockReturnValue({
      todo: mockTodoList,
      deleteTodo: mockDeleteTodo,
    });
  });

  it('renders the Home component correctly', () => {
    render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    // Check that the search input, status select, and TodoList component render
    expect(screen.getByLabelText('Search')).toBeInTheDocument();
    expect(screen.getByLabelText('Status')).toBeInTheDocument();
    expect(screen.getByText('To-Do List')).toBeInTheDocument();
    expect(screen.getByText('Add Todo')).toBeInTheDocument();
  });

  it('filters todos based on search input', () => {
    render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    const searchInput = screen.getByPlaceholderText(
      'Title'
    ) as HTMLInputElement;
    fireEvent.change(searchInput, { target: { value: 'First' } });

    // Check that only the 'First Todo' is displayed
    expect(screen.getByText('First Todo')).toBeInTheDocument();
    expect(screen.queryByText('Second Todo')).not.toBeInTheDocument();
    expect(screen.queryByText('Completed Todo')).not.toBeInTheDocument();
  });

  it('filters todos based on status', () => {
    render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    const statusSelect = screen.getByLabelText('Status') as HTMLSelectElement;
    fireEvent.change(statusSelect, { target: { value: 'Pending' } });

    // Check that only 'Pending' todos are displayed
    expect(screen.getByText('First Todo')).toBeInTheDocument();
    expect(screen.queryByText('Second Todo')).not.toBeInTheDocument();
    expect(screen.queryByText('Completed Todo')).not.toBeInTheDocument();
  });

  it('filters todos based on both search input and status filter', () => {
    render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    const searchInput = screen.getByPlaceholderText(
      'Title'
    ) as HTMLInputElement;
    const statusSelect = screen.getByLabelText('Status') as HTMLSelectElement;

    // Search for "Todo" and filter by "In Progress"
    fireEvent.change(searchInput, { target: { value: 'Todo' } });
    fireEvent.change(statusSelect, { target: { value: 'In Progress' } });

    // Ensure only the 'Second Todo' is shown
    expect(screen.queryByText('First Todo')).not.toBeInTheDocument();
    expect(screen.getByText('Second Todo')).toBeInTheDocument();
    expect(screen.queryByText('Completed Todo')).not.toBeInTheDocument();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    // Capture the snapshot of the component's rendered output
    expect(asFragment()).toMatchSnapshot();
  });
});

export {};
