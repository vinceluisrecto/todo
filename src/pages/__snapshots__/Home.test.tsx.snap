// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="min-w-full min-h-[100vh] flex items-center justify-center"
  >
    <div
      class="my-5 flex flex-col items-center justify-center w-full max-w-[35%] border border-gray-300 rounded-lg p-6 gap-3"
    >
      <h1
        class="text-[29px]"
      >
        To-Do List
      </h1>
      <div
        class="flex flex-row items-center justify-between w-full gap-3"
      >
        <div
          class="flex-1"
        >
          <label
            class="block mb-1 text-sm font-medium text-gray-900"
            for="search"
          >
            Search
          </label>
          <input
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
            id="search"
            name="title"
            placeholder="Title"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            class="block mb-1 text-sm font-medium text-gray-900"
            for="status"
          >
            Status
          </label>
          <select
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
            id="status"
            name="status"
          >
            <option
              value="All"
            >
              All
            </option>
            <option
              value="Pending"
            >
              Pending
            </option>
            <option
              value="In Progress"
            >
              In Progress
            </option>
            <option
              value="Completed"
            >
              Completed
            </option>
          </select>
        </div>
      </div>
      <div
        class="w-full"
      >
        <div
          class="w-full min-h-[200px]"
        >
          <div
            class="w-full my-1 py-2 border-b-[1px] border-[#d1d5db]"
          >
            <div
              class="flex flex-row items-center px-2 "
              data-testid="container-todo-item"
            >
              <div
                class="flex-1"
              >
                <span
                  class="text-[16px]"
                >
                  First Todo
                </span>
                <p
                  class="text-[11px] opacity-70"
                >
                  First Description
                </p>
                <p
                  class="text-[11px] opacity-70"
                >
                  Due: 2024-11-21
                </p>
              </div>
              <div
                class="flex items-center gap-3"
              >
                <span
                  class="inline-flex items-center rounded-md px-2 py-1 text-xs font-medium text-gray-600 ring-1 ring-inset ring-gray-500/10 bg-gray-50"
                >
                  Pending
                </span>
                <a
                  href="/add-todo?id=1"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-pen-to-square "
                    color="blue"
                    data-icon="pen-to-square"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M471.6 21.7c-21.9-21.9-57.3-21.9-79.2 0L362.3 51.7l97.9 97.9 30.1-30.1c21.9-21.9 21.9-57.3 0-79.2L471.6 21.7zm-299.2 220c-6.1 6.1-10.8 13.6-13.5 21.9l-29.6 88.8c-2.9 8.6-.6 18.1 5.8 24.6s15.9 8.7 24.6 5.8l88.8-29.6c8.2-2.7 15.7-7.4 21.9-13.5L437.7 172.3 339.7 74.3 172.4 241.7zM96 64C43 64 0 107 0 160L0 416c0 53 43 96 96 96l256 0c53 0 96-43 96-96l0-96c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 96c0 17.7-14.3 32-32 32L96 448c-17.7 0-32-14.3-32-32l0-256c0-17.7 14.3-32 32-32l96 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L96 64z"
                      fill="currentColor"
                    />
                  </svg>
                </a>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-trash cursor-pointer"
                  color="red"
                  data-icon="trash"
                  data-prefix="fas"
                  data-testid="delete-item"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M135.2 17.7L128 32 32 32C14.3 32 0 46.3 0 64S14.3 96 32 96l384 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-96 0-7.2-14.3C307.4 6.8 296.3 0 284.2 0L163.8 0c-12.1 0-23.2 6.8-28.6 17.7zM416 128L32 128 53.2 467c1.6 25.3 22.6 45 47.9 45l245.8 0c25.3 0 46.3-19.7 47.9-45L416 128z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="w-full my-1 py-2 border-b-[1px] border-[#d1d5db]"
          >
            <div
              class="flex flex-row items-center px-2 "
              data-testid="container-todo-item"
            >
              <div
                class="flex-1"
              >
                <span
                  class="text-[16px]"
                >
                  Second Todo
                </span>
                <p
                  class="text-[11px] opacity-70"
                >
                  Second Description
                </p>
                <p
                  class="text-[11px] opacity-70"
                >
                  Due: 2024-11-22
                </p>
              </div>
              <div
                class="flex items-center gap-3"
              >
                <span
                  class="inline-flex items-center rounded-md px-2 py-1 text-xs font-medium text-gray-600 ring-1 ring-inset ring-blue-700/10 bg-blue-50"
                >
                  In Progress
                </span>
                <a
                  href="/add-todo?id=2"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-pen-to-square "
                    color="blue"
                    data-icon="pen-to-square"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M471.6 21.7c-21.9-21.9-57.3-21.9-79.2 0L362.3 51.7l97.9 97.9 30.1-30.1c21.9-21.9 21.9-57.3 0-79.2L471.6 21.7zm-299.2 220c-6.1 6.1-10.8 13.6-13.5 21.9l-29.6 88.8c-2.9 8.6-.6 18.1 5.8 24.6s15.9 8.7 24.6 5.8l88.8-29.6c8.2-2.7 15.7-7.4 21.9-13.5L437.7 172.3 339.7 74.3 172.4 241.7zM96 64C43 64 0 107 0 160L0 416c0 53 43 96 96 96l256 0c53 0 96-43 96-96l0-96c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 96c0 17.7-14.3 32-32 32L96 448c-17.7 0-32-14.3-32-32l0-256c0-17.7 14.3-32 32-32l96 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L96 64z"
                      fill="currentColor"
                    />
                  </svg>
                </a>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-trash cursor-pointer"
                  color="red"
                  data-icon="trash"
                  data-prefix="fas"
                  data-testid="delete-item"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M135.2 17.7L128 32 32 32C14.3 32 0 46.3 0 64S14.3 96 32 96l384 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-96 0-7.2-14.3C307.4 6.8 296.3 0 284.2 0L163.8 0c-12.1 0-23.2 6.8-28.6 17.7zM416 128L32 128 53.2 467c1.6 25.3 22.6 45 47.9 45l245.8 0c25.3 0 46.3-19.7 47.9-45L416 128z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="w-full my-1 py-2 border-b-[1px] border-[#d1d5db]"
          >
            <div
              class="flex flex-row items-center px-2 "
              data-testid="container-todo-item"
            >
              <div
                class="flex-1"
              >
                <span
                  class="text-[16px]"
                >
                  Completed Todo
                </span>
                <p
                  class="text-[11px] opacity-70"
                >
                  Completed Description
                </p>
                <p
                  class="text-[11px] opacity-70"
                >
                  Due: 2024-11-23
                </p>
              </div>
              <div
                class="flex items-center gap-3"
              >
                <span
                  class="inline-flex items-center rounded-md px-2 py-1 text-xs font-medium text-gray-600 ring-1 ring-inset ring-green-600/20 bg-green-50"
                >
                  Completed
                </span>
                <a
                  href="/add-todo?id=3"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-pen-to-square "
                    color="blue"
                    data-icon="pen-to-square"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M471.6 21.7c-21.9-21.9-57.3-21.9-79.2 0L362.3 51.7l97.9 97.9 30.1-30.1c21.9-21.9 21.9-57.3 0-79.2L471.6 21.7zm-299.2 220c-6.1 6.1-10.8 13.6-13.5 21.9l-29.6 88.8c-2.9 8.6-.6 18.1 5.8 24.6s15.9 8.7 24.6 5.8l88.8-29.6c8.2-2.7 15.7-7.4 21.9-13.5L437.7 172.3 339.7 74.3 172.4 241.7zM96 64C43 64 0 107 0 160L0 416c0 53 43 96 96 96l256 0c53 0 96-43 96-96l0-96c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 96c0 17.7-14.3 32-32 32L96 448c-17.7 0-32-14.3-32-32l0-256c0-17.7 14.3-32 32-32l96 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L96 64z"
                      fill="currentColor"
                    />
                  </svg>
                </a>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-trash cursor-pointer"
                  color="red"
                  data-icon="trash"
                  data-prefix="fas"
                  data-testid="delete-item"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M135.2 17.7L128 32 32 32C14.3 32 0 46.3 0 64S14.3 96 32 96l384 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-96 0-7.2-14.3C307.4 6.8 296.3 0 284.2 0L163.8 0c-12.1 0-23.2 6.8-28.6 17.7zM416 128L32 128 53.2 467c1.6 25.3 22.6 45 47.9 45l245.8 0c25.3 0 46.3-19.7 47.9-45L416 128z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <a
        class="focus:outline-none text-white bg-green-700 hover:bg-green-800 focus:ring-4 focus:ring-green-300 font-medium rounded-lg text-sm px-5 py-2.5 me-2 mt-3 w-full text-center"
        data-testid="add-todo-button"
        href="/add-todo"
      >
        Add Todo
      </a>
    </div>
  </div>
</DocumentFragment>
`;
